// nf3/dist/_chunks/libs/exsolve.mjs
import e, { lstatSync as t, realpathSync as n, statSync as r } from "node:fs";
import { URL as i, fileURLToPath as a, pathToFileURL as o } from "node:url";
import s, { isAbsolute as c } from "node:path";
import l from "node:assert";
import u from "node:process";
import { format as d, inspect as f } from "node:util";
const p = `_http_agent._http_client._http_common._http_incoming._http_outgoing._http_server._stream_duplex._stream_passthrough._stream_readable._stream_transform._stream_wrap._stream_writable._tls_common._tls_wrap.assert.assert/strict.async_hooks.buffer.child_process.cluster.console.constants.crypto.dgram.diagnostics_channel.dns.dns/promises.domain.events.fs.fs/promises.http.http2.https.inspector.inspector/promises.module.net.os.path.path/posix.path/win32.perf_hooks.process.punycode.querystring.readline.readline/promises.repl.stream.stream/consumers.stream/promises.stream/web.string_decoder.sys.timers.timers/promises.tls.trace_events.tty.url.util.util/types.v8.vm.wasi.worker_threads.zlib`.split(`.`), m = /^([A-Z][a-z\d]*)+$/, h = /* @__PURE__ */ new Set([
	`string`,
	`function`,
	`number`,
	`object`,
	`Function`,
	`Object`,
	`boolean`,
	`bigint`,
	`symbol`
]), g = /* @__PURE__ */ new Map();
function _(e, t = `and`) {
	switch (e.length) {
		case 0: return ``;
		case 1: return `${e[0]}`;
		case 2: return `${e[0]} ${t} ${e[1]}`;
		case 3: return `${e[0]}, ${e[1]}, ${t} ${e[2]}`;
		default: return `${e.slice(0, -1).join(`, `)}, ${t} ${e.at(-1)}`;
	}
}
function v(e) {
	let t = 0, n = /%[dfijoOs]/g;
	for (; n.exec(e) !== null;) t++;
	return t;
}
function y(e, t, n) {
	return g.set(e, t), ee(n, e);
}
const b = Symbol(`kIsNodeError`);
function ee(e, t) {
	let n = g.get(t), r = typeof n == `string` ? v(n) : -1;
	switch (r) {
		case 0: {
			class i extends e {
				code = t;
				constructor(...e) {
					l.ok(e.length === 0, `Code: ${t}; The provided arguments length (${e.length}) does not match the required ones (${r}).`), super(n);
				}
				get ["constructor"]() {
					return e;
				}
				get [b]() {
					return !0;
				}
				toString() {
					return `${this.name} [${t}]: ${this.message}`;
				}
			}
			return i;
		}
		case -1: {
			class n extends e {
				code = t;
				constructor(...e) {
					super(), Object.defineProperty(this, "message", {
						value: te(t, e, this),
						enumerable: !1,
						writable: !0,
						configurable: !0
					});
				}
				get ["constructor"]() {
					return e;
				}
				get [b]() {
					return !0;
				}
				toString() {
					return `${this.name} [${t}]: ${this.message}`;
				}
			}
			return n;
		}
		default: {
			class i extends e {
				code = t;
				constructor(...e) {
					l.ok(e.length === r, `Code: ${t}; The provided arguments length (${e.length}) does not match the required ones (${r}).`), e.unshift(n), super(Reflect.apply(d, null, e));
				}
				get ["constructor"]() {
					return e;
				}
				get [b]() {
					return !0;
				}
				toString() {
					return `${this.name} [${t}]: ${this.message}`;
				}
			}
			return i;
		}
	}
}
function te(e, t, n) {
	let r = g.get(e);
	if (l.ok(r !== void 0, "expected `message` to be found"), typeof r == `function`) return l.ok(r.length <= t.length, `Code: ${e}; The provided arguments length (${t.length}) does not match the required ones (${r.length}).`), Reflect.apply(r, n, t);
	let i = v(r);
	return l.ok(i === t.length, `Code: ${e}; The provided arguments length (${t.length}) does not match the required ones (${i}).`), t.length === 0 ? r : (t.unshift(r), Reflect.apply(d, null, t));
}
function x(e) {
	if (e === null) return `null`;
	if (e === void 0) return `undefined`;
	let t = typeof e;
	switch (t) {
		case `bigint`: return `type bigint (${e}n)`;
		case `number`: return e === 0 ? 1 / e == -Infinity ? `type number (-0)` : `type number (0)` : Number.isNaN(e) ? `type number (NaN)` : e === Infinity ? `type number (Infinity)` : e === -Infinity ? `type number (-Infinity)` : `type number (${e})`;
		case `boolean`: return e ? `type boolean (true)` : `type boolean (false)`;
		case `symbol`: return `type symbol (${String(e)})`;
		case `function`: return `function ${e.name}`;
		case `object`: return e.constructor && e.constructor.name ? `an instance of ${e.constructor.name}` : `${f(e, { depth: -1 })}`;
		case `string`: {
			let t = e;
			return t.length > 28 && (t = `${t.slice(0, 25)}...`), t.includes(`'`) ? `type string (${JSON.stringify(t)})` : `type string ('${t}')`;
		}
		default: {
			let n = f(e, { colors: !1 });
			return n.length > 28 && (n = `${n.slice(0, 25)}...`), `type ${t} (${n})`;
		}
	}
}
y(`ERR_INVALID_ARG_TYPE`, (e, t, n) => {
	l.ok(typeof e == `string`, `'name' must be a string`), Array.isArray(t) || (t = [t]);
	let r = `The `;
	if (e.endsWith(` argument`)) r += `${e} `;
	else {
		let t = e.includes(`.`) ? `property` : `argument`;
		r += `"${e}" ${t} `;
	}
	r += `must be `;
	let i = [], a = [], o = [];
	for (let e of t) l.ok(typeof e == `string`, `All expected entries have to be of type string`), h.has(e) ? i.push(e.toLowerCase()) : m.exec(e) === null ? (l.ok(e !== `object`, `The value "object" should be written as "Object"`), o.push(e)) : a.push(e);
	if (a.length > 0) {
		let e = i.indexOf(`object`);
		e !== -1 && (i.splice(e, 1), a.push(`Object`));
	}
	return i.length > 0 && (r += `${i.length > 1 ? `one of type` : `of type`} ${_(i, `or`)}`, (a.length > 0 || o.length > 0) && (r += ` or `)), a.length > 0 && (r += `an instance of ${_(a, `or`)}`, o.length > 0 && (r += ` or `)), o.length > 0 && (o.length > 1 ? r += `one of ${_(o, `or`)}` : (o[0]?.toLowerCase() !== o[0] && (r += `an `), r += `${o[0]}`)), r += `. Received ${x(n)}`, r;
}, TypeError);
const S = y(`ERR_INVALID_MODULE_SPECIFIER`, (e, t, n) => `Invalid module "${e}" ${t}${n ? ` imported from ${n}` : ``}`, TypeError), C = y(`ERR_INVALID_PACKAGE_CONFIG`, (e, t, n) => `Invalid package config ${e}${t ? ` while importing ${t}` : ``}${n ? `. ${n}` : ``}`, Error), w = y(`ERR_INVALID_PACKAGE_TARGET`, (e, t, n, r = !1, i) => {
	let a = typeof n == `string` && !r && n.length > 0 && !n.startsWith(`./`);
	return t === `.` ? (l.ok(r === !1), `Invalid "exports" main target ${JSON.stringify(n)} defined in the package config ${e}package.json${i ? ` imported from ${i}` : ``}${a ? `; targets must start with "./"` : ``}`) : `Invalid "${r ? `imports` : `exports`}" target ${JSON.stringify(n)} defined for '${t}' in the package config ${e}package.json${i ? ` imported from ${i}` : ``}${a ? `; targets must start with "./"` : ``}`;
}, Error), T = y(`ERR_MODULE_NOT_FOUND`, function(e, t, n = !1) {
	return n && typeof n == `string` && (this.url = `${n}`), `Cannot find ${n ? `module` : `package`} '${e}' imported from ${t}`;
}, Error), E = y(`ERR_PACKAGE_IMPORT_NOT_DEFINED`, (e, t, n) => `Package import specifier "${e}" is not defined${t ? ` in package ${t || ``}package.json` : ``} imported from ${n}`, TypeError), ne = y(`ERR_PACKAGE_PATH_NOT_EXPORTED`, (e, t, n) => t === `.` ? `No "exports" main defined in ${e}package.json${n ? ` imported from ${n}` : ``}` : `Package subpath '${t}' is not defined by "exports" in ${e}package.json${n ? ` imported from ${n}` : ``}`, Error), re = y(`ERR_UNSUPPORTED_DIR_IMPORT`, function(e, t, n = void 0) {
	return this.url = n, `Directory import '${e}' is not supported resolving ES modules imported from ${t}`;
}, Error), D = y(`ERR_UNSUPPORTED_RESOLVE_REQUEST`, `Failed to resolve module specifier "%s" from "%s": Invalid relative URL or base scheme is not hierarchical.`, TypeError), ie = y(`ERR_UNKNOWN_FILE_EXTENSION`, (e, t) => `Unknown file extension "${e}" for ${t}`, TypeError);
y(`ERR_INVALID_ARG_VALUE`, (e, t, n = `is invalid`) => {
	let r = f(t);
	return r.length > 128 && (r = `${r.slice(0, 128)}...`), `The ${e.includes(`.`) ? `property` : `argument`} '${e}' ${n}. Received ${r}`;
}, TypeError);
const O = {}.hasOwnProperty, k = /* @__PURE__ */ new Map();
function A(t, { base: n, specifier: r }) {
	let i = k.get(t);
	if (i) return i;
	let o;
	try {
		o = e.readFileSync(s.toNamespacedPath(t), `utf8`);
	} catch (e) {
		let t = e;
		if (t.code !== `ENOENT`) throw t;
	}
	let c = {
		exists: !1,
		pjsonPath: t,
		main: void 0,
		name: void 0,
		type: `none`,
		exports: void 0,
		imports: void 0
	};
	if (o !== void 0) {
		let e;
		try {
			e = JSON.parse(o);
		} catch (e) {
			let i = new C(t, (n ? `"${r}" from ` : ``) + a(n || r), e.message);
			throw i.cause = e, i;
		}
		c.exists = !0, O.call(e, `name`) && typeof e.name == `string` && (c.name = e.name), O.call(e, `main`) && typeof e.main == `string` && (c.main = e.main), O.call(e, `exports`) && (c.exports = e.exports), O.call(e, `imports`) && (c.imports = e.imports), O.call(e, `type`) && (e.type === `commonjs` || e.type === `module`) && (c.type = e.type);
	}
	return k.set(t, c), c;
}
function j(e) {
	let t = new URL(`package.json`, e);
	for (; !t.pathname.endsWith(`node_modules/package.json`);) {
		let n = A(a(t), { specifier: e });
		if (n.exists) return n;
		let r = t;
		if (t = new URL(`../package.json`, t), t.pathname === r.pathname) break;
	}
	return {
		pjsonPath: a(t),
		exists: !1,
		type: `none`
	};
}
const ae = {}.hasOwnProperty, oe = {
	__proto__: null,
	".json": `json`,
	".cjs": `commonjs`,
	".cts": `commonjs`,
	".js": `module`,
	".ts": `module`,
	".mts": `module`,
	".mjs": `module`
}, M = {
	__proto__: null,
	"data:": ce,
	"file:": ue,
	"node:": () => `builtin`
};
function se(e) {
	return e && /^\s*(text|application)\/javascript\s*(;\s*charset=utf-?8\s*)?$/i.test(e) ? `module` : e === `application/json` ? `json` : null;
}
function ce(e) {
	let { 1: t } = /^([^/]+\/[^;,]+)(?:[^,]*?)(;base64)?,/.exec(e.pathname) || [
		null,
		null,
		null
	];
	return se(t);
}
function le(e) {
	let t = e.pathname;
	for (let e = t.length - 1; e > 0; e--) switch (t.charCodeAt(e)) {
		case 47: return ``;
		case 46: return t.charCodeAt(e - 1) === 47 ? `` : t.slice(e);
	}
	return ``;
}
function ue(e, t, n) {
	let r = le(e);
	if (r === `.js`) {
		let { type: t } = j(e);
		return t === `none` ? `commonjs` : t;
	}
	if (r === ``) {
		let { type: t } = j(e);
		return t === `module` ? `module` : t === `none` ? `commonjs` : t;
	}
	let i = oe[r];
	if (i) return i;
	if (!n) throw new ie(r, a(e));
}
function de(e, t) {
	let n = e.protocol;
	return ae.call(M, n) && M[n](e, t, !0) || null;
}
const N = RegExp.prototype[Symbol.replace], P = {}.hasOwnProperty, F = /(^|\\|\/)((\.|%2e)(\.|%2e)?|(n|%6e|%4e)(o|%6f|%4f)(d|%64|%44)(e|%65|%45)(_|%5f)(m|%6d|%4d)(o|%6f|%4f)(d|%64|%44)(u|%75|%55)(l|%6c|%4c)(e|%65|%45)(s|%73|%53))?(\\|\/|$)/i, I = /(^|\\|\/)((\.|%2e)(\.|%2e)?|(n|%6e|%4e)(o|%6f|%4f)(d|%64|%44)(e|%65|%45)(_|%5f)(m|%6d|%4d)(o|%6f|%4f)(d|%64|%44)(u|%75|%55)(l|%6c|%4c)(e|%65|%45)(s|%73|%53))(\\|\/|$)/i, fe = /^\.|%|\\/, L = /\*/g, R = /%2f|%5c/i, z = /* @__PURE__ */ new Set(), B = /[/\\]{2}/;
function V(e, t, n, r, i, o, s) {
	if (u.noDeprecation) return;
	let c = a(r), l = B.exec(s ? e : t) !== null;
	u.emitWarning(`Use of deprecated ${l ? `double slash` : `leading or trailing slash matching`} resolving "${e}" for module request "${t}" ${t === n ? `` : `matched to "${n}" `}in the "${i ? `imports` : `exports`}" field module resolution of the package at ${c}${o ? ` imported from ${a(o)}` : ``}.`, `DeprecationWarning`, `DEP0166`);
}
function H(e, t, n, r) {
	if (u.noDeprecation || de(e, { parentURL: n.href }) !== `module`) return;
	let o = a(e.href), c = a(new i(`.`, t)), l = a(n);
	r ? s.resolve(c, r) !== o && u.emitWarning(`Package ${c} has a "main" field set to "${r}", excluding the full filename and extension to the resolved file at "${o.slice(c.length)}", imported from ${l}.\n Automatic extension resolution of the "main" field is deprecated for ES modules.`, `DeprecationWarning`, `DEP0151`) : u.emitWarning(`No "main" or "exports" field defined in the package.json for ${c} resolving the main entry point "${o.slice(c.length)}", imported from ${l}.\nDefault "index" lookups for the main are deprecated for ES modules.`, `DeprecationWarning`, `DEP0151`);
}
function U(e) {
	try {
		return r(e);
	} catch {}
}
function W(e) {
	let t = r(e, { throwIfNoEntry: !1 });
	return (t ? t.isFile() : void 0) ?? !1;
}
function pe(e, t, n) {
	let r;
	if (t.main !== void 0) {
		if (r = new i(t.main, e), W(r)) return r;
		let a = [
			`./${t.main}.js`,
			`./${t.main}.json`,
			`./${t.main}.node`,
			`./${t.main}/index.js`,
			`./${t.main}/index.json`,
			`./${t.main}/index.node`
		], o = -1;
		for (; ++o < a.length && (r = new i(a[o], e), !W(r));) r = void 0;
		if (r) return H(r, e, n, t.main), r;
	}
	let o = [
		`./index.js`,
		`./index.json`,
		`./index.node`
	], s = -1;
	for (; ++s < o.length && (r = new i(o[s], e), !W(r));) r = void 0;
	if (r) return H(r, e, n, t.main), r;
	throw new T(a(new i(`.`, e)), a(n));
}
function me(e, t, r) {
	if (R.exec(e.pathname) !== null) throw new S(e.pathname, String.raw`must not include encoded "/" or "\" characters`, a(t));
	let i;
	try {
		i = a(e);
	} catch (n) {
		throw Object.defineProperty(n, "input", { value: String(e) }), Object.defineProperty(n, "module", { value: String(t) }), n;
	}
	let c = U(i.endsWith(`/`) ? i.slice(-1) : i);
	if (c && c.isDirectory()) {
		let n = new re(i, a(t));
		throw n.url = String(e), n;
	}
	if (!c || !c.isFile()) {
		let n = new T(i || e.pathname, t && a(t), !0);
		throw n.url = String(e), n;
	}
	if (!r) {
		let t = n(i), { search: r, hash: a } = e;
		e = o(t + (i.endsWith(s.sep) ? `/` : ``)), e.search = r, e.hash = a;
	}
	return e;
}
function he(e, t, n) {
	return new E(e, t && a(new i(`.`, t)), a(n));
}
function G(e, t, n) {
	return new ne(a(new i(`.`, t)), e, n && a(n));
}
function ge(e, t, n, r, i) {
	throw new S(e, `request is not a valid match in pattern "${t}" for the "${r ? `imports` : `exports`}" resolution of ${a(n)}`, i && a(i));
}
function K(e, t, n, r, o) {
	return t = typeof t == `object` && t ? JSON.stringify(t, null, ``) : `${t}`, new w(a(new i(`.`, n)), e, t, r, o && a(o));
}
function _e(e, t, n, r, a, o, s, c, l) {
	if (t !== `` && !o && e.at(-1) !== `/`) throw K(n, e, r, s, a);
	if (!e.startsWith(`./`)) {
		if (s && !e.startsWith(`../`) && !e.startsWith(`/`)) {
			let n = !1;
			try {
				new i(e), n = !0;
			} catch {}
			if (!n) return X(o ? N.call(L, e, () => t) : e + t, r, l);
		}
		throw K(n, e, r, s, a);
	}
	if (F.exec(e.slice(2)) !== null) if (I.exec(e.slice(2)) === null) {
		if (!c) {
			let i = o ? n.replace(`*`, () => t) : n + t;
			V(o ? N.call(L, e, () => t) : e, i, n, r, s, a, !0);
		}
	} else throw K(n, e, r, s, a);
	let u = new i(e, r), d = u.pathname, f = new i(`.`, r).pathname;
	if (!d.startsWith(f)) throw K(n, e, r, s, a);
	if (t === ``) return u;
	if (F.exec(t) !== null) {
		let i = o ? n.replace(`*`, () => t) : n + t;
		I.exec(t) === null ? c || V(o ? N.call(L, e, () => t) : e, i, n, r, s, a, !1) : ge(i, n, r, s, a);
	}
	return o ? new i(N.call(L, u.href, () => t)) : new i(t, u);
}
function ve(e) {
	let t = Number(e);
	return `${t}` === e ? t >= 0 && t < 4294967295 : !1;
}
function q(e, t, n, r, i, o, s, c, l) {
	if (typeof t == `string`) return _e(t, n, r, e, i, o, s, c, l);
	if (Array.isArray(t)) {
		let a = t;
		if (a.length === 0) return null;
		let u, d = -1;
		for (; ++d < a.length;) {
			let t = a[d], f;
			try {
				f = q(e, t, n, r, i, o, s, c, l);
			} catch (e) {
				let t = e;
				if (u = t, t.code === `ERR_INVALID_PACKAGE_TARGET`) continue;
				throw e;
			}
			if (f !== void 0) {
				if (f === null) {
					u = null;
					continue;
				}
				return f;
			}
		}
		if (u == null) return u;
		throw u;
	}
	if (typeof t == `object` && t) {
		let u = Object.getOwnPropertyNames(t), d = -1;
		for (; ++d < u.length;) {
			let t = u[d];
			if (ve(t)) throw new C(a(e), a(i), `"exports" cannot contain numeric property keys.`);
		}
		for (d = -1; ++d < u.length;) {
			let a = u[d];
			if (a === "default" || l && l.has(a)) {
				let u = t[a], d = q(e, u, n, r, i, o, s, c, l);
				if (d === void 0) continue;
				return d;
			}
		}
		return;
	}
	if (t === null) return null;
	throw K(r, t, e, s, i);
}
function ye(e, t, n) {
	if (typeof e == `string` || Array.isArray(e)) return !0;
	if (typeof e != `object` || !e) return !1;
	let r = Object.getOwnPropertyNames(e), i = !1, o = 0, s = -1;
	for (; ++s < r.length;) {
		let e = r[s], c = e === `` || e[0] !== `.`;
		if (o++ === 0) i = c;
		else if (i !== c) throw new C(a(t), a(n), `"exports" cannot contain some keys starting with '.' and some not. The exports object must either be an object of package subpath keys or an object of main entry condition name keys only.`);
	}
	return i;
}
function be(e, t, n) {
	if (u.noDeprecation) return;
	let r = a(t);
	z.has(r + `|` + e) || (z.add(r + `|` + e), u.emitWarning(`Use of deprecated trailing slash pattern mapping "${e}" in the "exports" field module resolution of the package at ${r}${n ? ` imported from ${a(n)}` : ``}. Mapping specifiers ending in "/" is no longer supported.`, `DeprecationWarning`, `DEP0155`));
}
function J(e, t, n, r, i) {
	let a = n.exports;
	if (ye(a, e, r) && (a = { ".": a }), P.call(a, t) && !t.includes(`*`) && !t.endsWith(`/`)) {
		let n = a[t], o = q(e, n, ``, t, r, !1, !1, !1, i);
		if (o == null) throw G(t, e, r);
		return o;
	}
	let o = ``, s = ``, c = Object.getOwnPropertyNames(a), l = -1;
	for (; ++l < c.length;) {
		let n = c[l], i = n.indexOf(`*`);
		if (i !== -1 && t.startsWith(n.slice(0, i))) {
			t.endsWith(`/`) && be(t, e, r);
			let a = n.slice(i + 1);
			t.length >= n.length && t.endsWith(a) && Y(o, n) === 1 && n.lastIndexOf(`*`) === i && (o = n, s = t.slice(i, t.length - a.length));
		}
	}
	if (o) {
		let n = a[o], c = q(e, n, s, o, r, !0, !1, t.endsWith(`/`), i);
		if (c == null) throw G(t, e, r);
		return c;
	}
	throw G(t, e, r);
}
function Y(e, t) {
	let n = e.indexOf(`*`), r = t.indexOf(`*`), i = n === -1 ? e.length : n + 1, a = r === -1 ? t.length : r + 1;
	return i > a ? -1 : a > i || n === -1 ? 1 : r === -1 || e.length > t.length ? -1 : +(t.length > e.length);
}
function xe(e, t, n) {
	if (e === `#` || e.endsWith(`/`)) throw new S(e, `is not a valid internal imports specifier name`, a(t));
	let r, i = j(t);
	if (i.exists) {
		r = o(i.pjsonPath);
		let a = i.imports;
		if (a) if (P.call(a, e) && !e.includes(`*`)) {
			let i = q(r, a[e], ``, e, t, !1, !0, !1, n);
			if (i != null) return i;
		} else {
			let i = ``, o = ``, s = Object.getOwnPropertyNames(a), c = -1;
			for (; ++c < s.length;) {
				let t = s[c], n = t.indexOf(`*`);
				if (n !== -1 && e.startsWith(t.slice(0, -1))) {
					let r = t.slice(n + 1);
					e.length >= t.length && e.endsWith(r) && Y(i, t) === 1 && t.lastIndexOf(`*`) === n && (i = t, o = e.slice(n, e.length - r.length));
				}
			}
			if (i) {
				let e = a[i], s = q(r, e, o, i, t, !0, !0, !1, n);
				if (s != null) return s;
			}
		}
	}
	throw he(e, r, t);
}
function Se(e, t) {
	let n = e.indexOf(`/`), r = !0, i = !1;
	e[0] === `@` && (i = !0, n === -1 || e.length === 0 ? r = !1 : n = e.indexOf(`/`, n + 1));
	let o = n === -1 ? e : e.slice(0, n);
	if (fe.exec(o) !== null && (r = !1), !r) throw new S(e, `is not a valid package name`, a(t));
	return {
		packageName: o,
		packageSubpath: `.` + (n === -1 ? `` : e.slice(n)),
		isScoped: i
	};
}
function X(e, t, n) {
	if (p.includes(e)) return new i(`node:` + e);
	let { packageName: r, packageSubpath: s, isScoped: c } = Se(e, t), l = j(t);
	if (l.exists && l.name === r && l.exports !== void 0 && l.exports !== null) return J(o(l.pjsonPath), s, l, t, n);
	let u = new i(`./node_modules/` + r + `/package.json`, t), d = a(u), f;
	do {
		let o = U(d.slice(0, -13));
		if (!o || !o.isDirectory()) {
			f = d, u = new i((c ? `../../../../node_modules/` : `../../../node_modules/`) + r + `/package.json`, u), d = a(u);
			continue;
		}
		let l = A(d, {
			base: t,
			specifier: e
		});
		return l.exports !== void 0 && l.exports !== null ? J(u, s, l, t, n) : s === `.` ? pe(u, l, t) : new i(s, u);
	} while (d.length !== f.length);
	throw new T(r, a(t), !1);
}
function Ce(e) {
	return e[0] === `.` && (e.length === 1 || e[1] === `/` || e[1] === `.` && (e.length === 2 || e[2] === `/`));
}
function we(e) {
	return e === `` ? !1 : e[0] === `/` ? !0 : Ce(e);
}
function Te(e, t, n, r) {
	let a = t.protocol, o = a === `data:`, s;
	if (we(e)) try {
		s = new i(e, t);
	} catch (n) {
		let r = new D(e, t);
		throw r.cause = n, r;
	}
	else if (a === `file:` && e[0] === `#`) s = xe(e, t, n);
	else try {
		s = new i(e);
	} catch (r) {
		if (o && !p.includes(e)) {
			let n = new D(e, t);
			throw n.cause = r, n;
		}
		s = X(e, t, n);
	}
	return l.ok(s !== void 0, `expected to be defined`), s.protocol === `file:` ? me(s, t, r) : s;
}
const Ee = /* @__PURE__ */ new Set([`node`, `import`]), De = process.platform === `win32`, Oe = globalThis.__EXSOLVE_CACHE__ ||= /* @__PURE__ */ new Map();
function ke(e, r) {
	let i = Re(e);
	if (`external` in i) return i.external;
	let a = i.specifier, s = i.url, c = i.absolutePath, l, u;
	if (r?.cache !== !1 && (l = Pe(c || a, r), u = r?.cache && typeof r?.cache == `object` ? r.cache : Oe), u) {
		let e = u.get(l);
		if (typeof e == `string`) return e;
		if (e instanceof Error) {
			if (r?.try) return;
			throw e;
		}
	}
	if (c) try {
		let e = t(c);
		if (e.isSymbolicLink() && (c = n(c), s = o(c)), e.isFile()) return u && u.set(l, s.href), s.href;
	} catch (e) {
		if (e?.code !== `ENOENT`) throw u && u.set(l, e), e;
	}
	let d = r?.conditions ? new Set(r.conditions) : Ee, f = a || s.href, p = je(r?.from), m = r?.suffixes || [``], h = r?.extensions ? [``, ...r.extensions] : [``], g;
	for (let e of p) {
		for (let t of m) {
			let n = Ie(f, t);
			n === `.` && (n += `/.`);
			for (let t of h) if (g = Ae(n + t, e, d), g) break;
			if (g) break;
		}
		if (g) break;
	}
	if (!g) {
		let t = Error(`Cannot resolve module "${e}" (from: ${p.map((e) => Ne(e)).join(`, `)})`);
		if (t.code = `ERR_MODULE_NOT_FOUND`, u && u.set(l, t), r?.try) return;
		throw t;
	}
	return u && u.set(l, g.href), g.href;
}
function Z(e, t) {
	let n = ke(e, t);
	if (!n || !n.startsWith(`file://`) && t?.try) return;
	let r = a(n);
	return De ? Le(r) : r;
}
function Ae(e, t, n) {
	try {
		return Te(e, t, n);
	} catch {}
}
function je(e) {
	let t = (Array.isArray(e) ? e : [e]).flatMap((e) => Me(e));
	return t.length === 0 ? [o(`./`)] : t;
}
function Me(e) {
	if (!e) return [];
	if ($(e)) return [e];
	if (typeof e != `string`) return [];
	if (/^(?:node|data|http|https|file):/.test(e)) return new URL(e);
	try {
		return e.endsWith(`/`) || r(e).isDirectory() ? o(e + `/`) : o(e);
	} catch {
		return [o(e + `/`), o(e)];
	}
}
function Ne(e) {
	try {
		return a(e);
	} catch {
		return e;
	}
}
function Pe(e, t) {
	let n;
	return Array.isArray(t?.from) ? n = t.from : t?.from && (n = [t.from]), Q([e]) + Fe(t?.conditions) + Q(t?.extensions) + Q(n) + Q(t?.suffixes);
}
function Fe(e) {
	return e ? Q([...new Set(e)].sort()) : `2:6:import4:node`;
}
function Q(e) {
	if (!e) return `-`;
	let t = `${e.length}:`;
	for (let n of e) {
		let e = String(n);
		t += `${e.length}:${e}`;
	}
	return t;
}
function Ie(e, t) {
	return !e || !t || t === `/` ? e : (e.endsWith(`/`) ? e : e + `/`) + (t.startsWith(`/`) ? t.slice(1) : t);
}
function Le(e) {
	return e.replace(/\\/g, `/`).replace(/^[a-z]:\//, (e) => e.toUpperCase());
}
function $(e) {
	return e instanceof URL || e?.constructor?.name === `URL`;
}
function Re(e) {
	if (typeof e == `string`) {
		if (e.startsWith(`file:`)) {
			let t = new URL(e);
			return {
				url: t,
				absolutePath: a(t)
			};
		}
		return c(e) ? {
			url: o(e),
			absolutePath: e
		} : /^(?:node|data|http|https):/.test(e) ? { external: e } : p.includes(e) && !e.includes(`:`) ? { external: `node:${e}` } : { specifier: e };
	}
	if ($(e)) return e.protocol === `file:` ? {
		url: e,
		absolutePath: a(e)
	} : { external: e.href };
	throw TypeError("id must be a `string` or `URL`");
}
export { Z as resolveModulePath };
