// @lovable.dev/vite-tanstack-config/dist/index.js
// src/index.ts
import { componentTagger } from "lovable-tagger";
import { mergeConfig, loadEnv } from "vite";

// src/assetsProxyPlugin.ts
import https from "https";
var ASSET_RE = /^\/__l5e\/assets-v1\//;
var PREVIEW_HOST_ENV = "LOVABLE_PREVIEW_HOST";
var HOP_BY_HOP = /* @__PURE__ */ new Set([
  "host",
  "connection",
  "keep-alive",
  "proxy-authenticate",
  "proxy-authorization",
  "te",
  "trailer",
  "transfer-encoding",
  "upgrade",
  "content-length",
  "cookie"
]);
function pickRequestHeaders(src) {
  const out = {};
  for (const [k, v] of Object.entries(src)) {
    if (v === void 0)
      continue;
    if (HOP_BY_HOP.has(k.toLowerCase()))
      continue;
    out[k] = v;
  }
  return out;
}
function lovableAssetsProxyPlugin() {
  const previewHost = process.env[PREVIEW_HOST_ENV]?.trim();
  return {
    name: "lovable:assets-proxy",
    apply: "serve",
    configureServer(server) {
      if (!previewHost)
        return;
      const middleware = (req, res, next) => {
        const url = req.url ?? "";
        if (!ASSET_RE.test(url))
          return next();
        const upstream = https.request(
          {
            host: previewHost,
            port: 443,
            method: req.method,
            path: url,
            headers: { ...pickRequestHeaders(req.headers), host: previewHost }
          },
          (upRes) => {
            res.statusCode = upRes.statusCode ?? 502;
            for (const [k, v] of Object.entries(upRes.headers)) {
              if (v === void 0)
                continue;
              if (HOP_BY_HOP.has(k.toLowerCase()))
                continue;
              res.setHeader(k, v);
            }
            upRes.pipe(res);
          }
        );
        upstream.on("error", (err) => {
          server.config.logger.warn(
            `[@lovable.dev/vite-tanstack-config] asset proxy ${url} \u2192 ${previewHost} failed: ${err.message}`
          );
          if (!res.headersSent) {
            res.statusCode = 502;
            res.end();
          } else {
            res.destroy(err);
          }
        });
        req.pipe(upstream);
      };
      server.middlewares.use(middleware);
    }
  };
}

// src/index.ts
var SANDBOX_ENV_VAR = "DEV_SERVER__PROJECT_PATH";
var LOVABLE_SANDBOX_ENV_VAR = "LOVABLE_SANDBOX";
function isSandboxEnvironment() {
  return process.env[LOVABLE_SANDBOX_ENV_VAR] === "1" || !!process.env[SANDBOX_ENV_VAR];
}
var TOOL_MANAGED_WATCH_IGNORES = [
  "**/.workspace/**",
  "**/.agents/**",
  "**/.claude/**",
  "**/.lovable/**",
  "**/.tanstack/tmp/**"
];
function logWarning(message) {
  console.warn(`[@lovable.dev/vite-tanstack-config] ${message}`);
}
var NITRO_DEFAULT_PRESET_MIN = [3, 0, 260603];
async function nitroSupportsDefaultPreset() {
  try {
    const { createRequire } = await import("module");
    const require2 = createRequire(`${process.cwd()}/package.json`);
    const { version } = require2("nitro/package.json");
    const match = typeof version === "string" ? /^(\d+)\.(\d+)\.(\d+)/.exec(version) : null;
    if (!match)
      return true;
    const parts = [Number(match[1]), Number(match[2]), Number(match[3])];
    for (let i = 0; i < 3; i++) {
      if (parts[i] !== NITRO_DEFAULT_PRESET_MIN[i])
        return parts[i] > NITRO_DEFAULT_PRESET_MIN[i];
    }
    return true;
  } catch {
    return true;
  }
}
function validateHeaders(config) {
  const headers = config.server?.headers;
  if (!headers)
    return;
  const headerKeys = Object.keys(headers);
  if (headerKeys.length > 0) {
    logWarning(`Your config includes server headers that will be removed in sandbox mode: ${headerKeys.join(", ")}`);
  }
}
function isExternalUrl(url) {
  try {
    const parsed = new URL(url);
    return !["localhost", "127.0.0.1", "::1", "0.0.0.0", "::"].includes(parsed.hostname);
  } catch {
    return false;
  }
}
function validateProxy(config) {
  const proxy = config.server?.proxy;
  if (!proxy)
    return;
  for (const [path, proxyConfig] of Object.entries(proxy)) {
    if (typeof proxyConfig === "string") {
      if (isExternalUrl(proxyConfig)) {
        logWarning(
          `Proxy '${path}' targets external URL: ${proxyConfig}. This may cause issues in the sandbox environment.`
        );
      }
    } else if (proxyConfig) {
      const target = typeof proxyConfig.target === "string" ? proxyConfig.target : void 0;
      if (target && isExternalUrl(target)) {
        logWarning(
          `Proxy '${path}' targets external URL: ${target}. This may cause issues in the sandbox environment.`
        );
      }
      if (proxyConfig.ws) {
        logWarning(`Proxy '${path}' enables WebSocket proxying, which may interfere with HMR.`);
      }
      if (proxyConfig.changeOrigin) {
        logWarning(`Proxy '${path}' changes origin, which may cause CORS issues.`);
      }
    }
  }
}
function validateCors(config) {
  if (config.server?.cors && typeof config.server.cors !== "boolean" && config.server.cors !== null) {
    logWarning("Custom CORS config detected. Sandbox requires specific CORS settings for proxy compatibility.");
  }
}
function validatePort(config) {
  if (config.server?.port && config.server.port !== 8080) {
    logWarning(`Port ${config.server.port} configured, but sandbox requires port 8080. Using 8080.`);
  }
}
function validateConfig(config) {
  validateHeaders(config);
  validateProxy(config);
  validateCors(config);
  validatePort(config);
}
function cleanServerConfig(config) {
  if (config.server?.headers || config.server?.cors || config.server?.proxy) {
    const { headers: _headers, cors: _cors, proxy: _proxy, ...serverRest } = config.server;
    return {
      ...config,
      server: serverRest
    };
  }
  return config;
}
function applyWatchDebounceDefaults(config) {
  const existingWatch = config.server?.watch ?? {};
  const existingAwaitWriteFinish = existingWatch.awaitWriteFinish;
  const hasAwaitWriteFinishObject = !!existingAwaitWriteFinish && typeof existingAwaitWriteFinish === "object" && !Array.isArray(existingAwaitWriteFinish);
  return mergeConfig(config, {
    server: {
      watch: {
        ...existingWatch,
        awaitWriteFinish: {
          ...hasAwaitWriteFinishObject ? existingAwaitWriteFinish : {},
          stabilityThreshold: 1e3,
          pollInterval: 100
        }
      }
    }
  });
}
var SSR_CAPTURE_KEY = "__LOVABLE_TANSTACK_CAPTURE_SSR_ERROR__";
var HANDLER_RAN_KEY = "__LOVABLE_TANSTACK_HANDLER_RAN__";
var BUILD_ERROR_MESSAGE = /transform failed with \d+ error|failed to parse source/i;
function isBuildError(error) {
  if (typeof error === "string")
    return BUILD_ERROR_MESSAGE.test(error);
  if (!error || typeof error !== "object")
    return false;
  const e = error;
  if (typeof e.plugin === "string" && (e.plugin.startsWith("vite:") || e.plugin.startsWith("esbuild:")))
    return true;
  if (e.frame != null && e.loc != null)
    return true;
  if (Array.isArray(e.errors) && e.errors.length > 0 && e.errors.every((x) => x != null && typeof x === "object" && "text" in x && "location" in x)) {
    return true;
  }
  const message = typeof e.message === "string" ? e.message : "";
  return BUILD_ERROR_MESSAGE.test(message);
}
var BUILD_DIAGNOSTICS_PLUGIN_NAME = "lovable-build-error-diagnostics";
var MAX_DIAGNOSTIC_JSON_LENGTH = 6e3;
var MAX_DIAGNOSTIC_STRING_LENGTH = 2e3;
var MAX_DIAGNOSTIC_NESTED_ERRORS = 20;
var MAX_DIAGNOSTIC_DEPTH = 4;
var COMPACT_DIAGNOSTIC_STRING_LENGTH = 300;
var COMPACT_DIAGNOSTIC_NESTED_ERRORS = 3;
var COMPACT_DIAGNOSTIC_DEPTH = 3;
var DIAGNOSTIC_ERROR_FIELDS = [
  "name",
  "message",
  "plugin",
  "pluginCode",
  "id",
  "moduleId",
  "hook",
  "code",
  "loc",
  "frame",
  "stack",
  "cause"
];
function truncateDiagnosticString(value) {
  return truncateDiagnosticStringTo(value, MAX_DIAGNOSTIC_STRING_LENGTH);
}
function truncateDiagnosticStringTo(value, maxLength) {
  return value.length > maxLength ? `${value.slice(0, maxLength)}...[truncated]` : value;
}
function readDiagnosticProperty(value, key) {
  try {
    return value[key];
  } catch (error) {
    return `[unreadable: ${error instanceof Error ? error.message : String(error)}]`;
  }
}
function serializeOwnDiagnosticProperties(value, depth, seen) {
  const keys = Object.keys(value);
  const output = {};
  for (const key of keys.slice(0, MAX_DIAGNOSTIC_NESTED_ERRORS)) {
    const nested = readDiagnosticProperty(value, key);
    if (nested == null || nested === "")
      continue;
    output[key] = serializeDiagnosticValue(nested, depth + 1, seen);
  }
  if (keys.length > MAX_DIAGNOSTIC_NESTED_ERRORS) {
    output.properties_truncated = keys.length - MAX_DIAGNOSTIC_NESTED_ERRORS;
  }
  return output;
}
function describeDiagnosticPrimitive(value) {
  if (typeof value === "symbol")
    return value.description ? `Symbol(${value.description})` : "Symbol()";
  if (typeof value === "bigint")
    return value.toString();
  if (typeof value === "function")
    return `[function ${value.name || "anonymous"}]`;
  return String(value);
}
function describeDiagnosticObject(value) {
  const constructorName = value.constructor?.name;
  if (typeof constructorName === "string" && constructorName !== "" && constructorName !== "Object") {
    return `[${constructorName}]`;
  }
  return Object.prototype.toString.call(value);
}
function serializeDiagnosticValue(value, depth, seen) {
  if (value == null)
    return value;
  if (typeof value === "string")
    return truncateDiagnosticString(value);
  if (typeof value === "number" || typeof value === "boolean")
    return value;
  if (typeof value !== "object")
    return describeDiagnosticPrimitive(value);
  if (seen.has(value))
    return "[circular]";
  if (depth >= MAX_DIAGNOSTIC_DEPTH)
    return "[max depth]";
  seen.add(value);
  try {
    if (Array.isArray(value)) {
      return value.slice(0, MAX_DIAGNOSTIC_NESTED_ERRORS).map((item) => serializeDiagnosticValue(item, depth + 1, seen));
    }
    const record = value;
    const output = {};
    for (const field of DIAGNOSTIC_ERROR_FIELDS) {
      if (!(field in record))
        continue;
      const nested = readDiagnosticProperty(record, field);
      if (nested == null || nested === "")
        continue;
      output[field] = serializeDiagnosticValue(nested, depth + 1, seen);
    }
    const nestedErrors = readDiagnosticProperty(record, "errors");
    if (Array.isArray(nestedErrors)) {
      output.errors = nestedErrors.slice(0, MAX_DIAGNOSTIC_NESTED_ERRORS).map((item) => serializeDiagnosticValue(item, depth + 1, seen));
      if (nestedErrors.length > MAX_DIAGNOSTIC_NESTED_ERRORS) {
        output.errors_truncated = nestedErrors.length - MAX_DIAGNOSTIC_NESTED_ERRORS;
      }
    } else if (nestedErrors != null && nestedErrors !== "") {
      output.errors = serializeDiagnosticValue(nestedErrors, depth + 1, seen);
    }
    if (Object.keys(output).length > 0)
      return output;
    const ownProperties = serializeOwnDiagnosticProperties(record, depth, seen);
    return Object.keys(ownProperties).length > 0 ? ownProperties : describeDiagnosticObject(value);
  } catch (error) {
    return `[unreadable object: ${error instanceof Error ? error.message : String(error)}]`;
  } finally {
    seen.delete(value);
  }
}
function compactDiagnosticValue(value, depth) {
  if (value == null)
    return value;
  if (typeof value === "string")
    return truncateDiagnosticStringTo(value, COMPACT_DIAGNOSTIC_STRING_LENGTH);
  if (typeof value === "number" || typeof value === "boolean")
    return value;
  if (typeof value !== "object")
    return describeDiagnosticPrimitive(value);
  if (depth >= COMPACT_DIAGNOSTIC_DEPTH)
    return "[max depth]";
  try {
    if (Array.isArray(value)) {
      const items = value.slice(0, COMPACT_DIAGNOSTIC_NESTED_ERRORS).map((item) => compactDiagnosticValue(item, depth + 1));
      if (value.length > COMPACT_DIAGNOSTIC_NESTED_ERRORS) {
        items.push(`[${value.length - COMPACT_DIAGNOSTIC_NESTED_ERRORS} items truncated]`);
      }
      return items;
    }
    const record = value;
    const output = {};
    const keys = Object.keys(record);
    for (const key of keys.slice(0, MAX_DIAGNOSTIC_NESTED_ERRORS)) {
      const nested = readDiagnosticProperty(record, key);
      if (nested == null || nested === "")
        continue;
      output[key] = compactDiagnosticValue(nested, depth + 1);
    }
    if (keys.length > MAX_DIAGNOSTIC_NESTED_ERRORS) {
      output.properties_truncated = keys.length - MAX_DIAGNOSTIC_NESTED_ERRORS;
    }
    return output;
  } catch (error) {
    return `[unreadable object: ${error instanceof Error ? error.message : String(error)}]`;
  }
}
function stringifyDiagnosticPayload(diagnostics) {
  const full = JSON.stringify(diagnostics);
  if (full && full.length <= MAX_DIAGNOSTIC_JSON_LENGTH)
    return full;
  const compact = compactDiagnosticValue(diagnostics, 0);
  const compactPayload = compact != null && typeof compact === "object" && !Array.isArray(compact) ? { ...compact, diagnostics_truncated: true } : { diagnostics_truncated: true, value: compact };
  const compactJson = JSON.stringify(compactPayload);
  if (compactJson.length <= MAX_DIAGNOSTIC_JSON_LENGTH)
    return compactJson;
  return JSON.stringify({
    diagnostics_truncated: true,
    message: "Diagnostic payload exceeded output budget."
  });
}
function lovableBuildErrorDiagnostics() {
  return {
    name: BUILD_DIAGNOSTICS_PLUGIN_NAME,
    apply: "build",
    buildEnd(error) {
      if (!error)
        return;
      try {
        const diagnostics = serializeDiagnosticValue(error, 0, /* @__PURE__ */ new WeakSet());
        const output = `
[lovable-build-error-diagnostics]
${stringifyDiagnosticPayload(diagnostics)}
[/lovable-build-error-diagnostics]
`;
        let written = false;
        const write = () => {
          if (written)
            return;
          written = true;
          try {
            process.stderr.write(output);
          } catch {
          }
        };
        setImmediate(write);
        process.once("exit", write);
      } catch {
      }
    }
  };
}
function devSsrErrorLogger() {
  let lastCapture;
  const CAPTURE_TTL_MS = 5e3;
  const handlerRanFlag = Symbol("lovableHandlerRan");
  const capture = (error) => {
    lastCapture = { error, at: Date.now() };
  };
  const consumeCapture = () => {
    if (!lastCapture)
      return void 0;
    if (Date.now() - lastCapture.at > CAPTURE_TTL_MS) {
      lastCapture = void 0;
      return void 0;
    }
    const { error } = lastCapture;
    lastCapture = void 0;
    return error;
  };
  return {
    name: "dev-ssr-error-logger",
    apply: "serve",
    configureServer(server) {
      globalThis[SSR_CAPTURE_KEY] = capture;
      globalThis[HANDLER_RAN_KEY] = (event) => {
        const res = event?.node?.res;
        if (res)
          res[handlerRanFlag] = true;
      };
      const g = globalThis;
      if (typeof g.addEventListener === "function") {
        const addListener = g.addEventListener;
        addListener("error", (e) => capture(e.error ?? e));
        addListener("unhandledrejection", (e) => capture(e.reason));
      }
      const onUnhandledRejection = (reason) => capture(reason);
      process.on("unhandledRejection", onUnhandledRejection);
      server.httpServer?.once("close", () => {
        process.off("unhandledRejection", onUnhandledRejection);
      });
      server.middlewares.use((req, res, next) => {
        let handled = false;
        const handleErrorResponse = () => {
          if (handled || res.statusCode < 500)
            return;
          handled = true;
          const captured = consumeCapture();
          const ranHandler = Boolean(res[handlerRanFlag]);
          let source;
          if (!ranHandler) {
            source = "vite";
          } else if (isBuildError(captured)) {
            source = "build";
          } else {
            source = "app";
          }
          if (!res.headersSent) {
            try {
              res.setHeader("x-lovable-error-source", source);
            } catch {
            }
          }
          const where = [req.method, req.url].filter(Boolean).join(" ") || "the request";
          let data;
          if (captured instanceof Error) {
            data = { name: captured.name, message: captured.message, stack: captured.stack };
          } else if (typeof captured === "string" && captured.length > 0) {
            data = { name: "Error", message: captured };
          } else if (ranHandler) {
            data = {
              message: `The app returned ${res.statusCode} while handling ${where}. The error was handled by a route or error boundary, so no stack was captured \u2014 check the failing loader/route code and the dev server output.`
            };
          } else {
            data = {
              message: `Dev server returned ${res.statusCode} for ${where} before the app handler ran. This is usually a Vite build/transform error \u2014 check the dev server output for the underlying error.`
            };
          }
          try {
            server.ws.send({ type: "custom", event: "server-ssr-error", data });
          } catch {
          }
        };
        const origWriteHead = res.writeHead.bind(res);
        res.writeHead = (...args) => {
          if (typeof args[0] === "number")
            res.statusCode = args[0];
          handleErrorResponse();
          return origWriteHead(...args);
        };
        const origEnd = res.end.bind(res);
        res.end = (...args) => {
          handleErrorResponse();
          return origEnd(...args);
        };
        next();
      });
    },
    transform(code, id) {
      const normalizedId = id.replace(/\\/g, "/");
      const isTargetModule = normalizedId.includes("/@tanstack/start-server-core/src/request-response.ts") || normalizedId.includes("/@tanstack/start-server-core/dist/esm/request-response.js");
      if (!isTargetModule) {
        return null;
      }
      const needle = "handler(request, requestOpts)";
      if (!code.includes(needle)) {
        return null;
      }
      return code.replace(
        needle,
        `(globalThis.${HANDLER_RAN_KEY}?.(typeof h3Event === "undefined" ? undefined : h3Event), Promise.resolve(${needle}).catch((err) => { globalThis.${SSR_CAPTURE_KEY}?.(err); throw err; }))`
      );
    }
  };
}
function devServerFnErrorLogger() {
  const HMR_SEND_KEY = "__TANSTACK_SERVER_FN_HMR_SEND__";
  return {
    name: "dev-server-fn-error-logger",
    apply: "serve",
    enforce: "pre",
    configureServer(server) {
      globalThis[HMR_SEND_KEY] = (data) => {
        server.ws.send({
          type: "custom",
          event: "server-fn-error",
          data
        });
      };
    },
    transform(code, id) {
      const normalizedId = id.replace(/\\/g, "/");
      const isTargetModule = normalizedId.includes("/@tanstack/start-server-core/src/server-functions-handler.ts") || normalizedId.includes("/@tanstack/start-server-core/dist/esm/server-functions-handler.js");
      if (!isTargetModule) {
        return null;
      }
      const needle = "const unwrapped = res.result || res.error";
      if (!code.includes(needle)) {
        return null;
      }
      return code.replace(
        needle,
        `${needle}

      if (res?.error) {
        const err = res.error
        const payload = {
          source: 'tanstack',
          type: 'server-fn-error',
          method: request.method,
          url: request.url,
          name: err?.name ?? 'Error',
          message: err?.message ?? String(err),
          stack: typeof err?.stack === 'string' ? err.stack : undefined,
        }
        globalThis.${HMR_SEND_KEY}?.(payload)
      }`
      );
    }
  };
}
function defineConfig(configOrOptions = {}) {
  return async (env) => {
    const { command, mode } = env;
    const isSandbox = isSandboxEnvironment();
    const isDevBuild = command === "build" && mode === "development";
    let options;
    if (typeof configOrOptions === "function") {
      const userConfig = await Promise.resolve(configOrOptions(env));
      options = { vite: userConfig };
    } else if (configOrOptions instanceof Promise) {
      options = { vite: await configOrOptions };
    } else {
      const optionObject = configOrOptions && typeof configOrOptions === "object" ? configOrOptions : {};
      if ("cloudflare" in optionObject) {
        if (optionObject.cloudflare === false && !("nitro" in optionObject)) {
          optionObject.nitro = false;
          logWarning(
            "`cloudflare: false` is deprecated in v2 \u2014 migrated to `nitro: false`. Update your config to drop the warning."
          );
        } else {
          logWarning(
            "The `cloudflare` option was removed in v2. Use `nitro: false` to disable the deploy plugin, or `nitro: { preset: '...' }` to configure it. The `cloudflare` setting is being ignored."
          );
        }
      }
      const hasLovableKey = "vite" in optionObject || "serverFnErrorLogger" in optionObject || "ssrErrorLogger" in optionObject || "nitro" in optionObject || "tanstackStart" in optionObject || "react" in optionObject || "envDefine" in optionObject || "hmrGate" in optionObject;
      options = hasLovableKey ? optionObject : { vite: optionObject };
    }
    const internalPlugins = [];
    const tailwindcss = (await import("@tailwindcss/vite")).default;
    internalPlugins.push(tailwindcss());
    const tsConfigPaths = (await import("vite-tsconfig-paths")).default;
    internalPlugins.push(tsConfigPaths({ projects: ["./tsconfig.json"] }));
    if (options.serverFnErrorLogger !== false) {
      internalPlugins.push(devServerFnErrorLogger());
    }
    if (options.ssrErrorLogger !== false) {
      internalPlugins.push(devSsrErrorLogger());
    }
    if (command === "build" && isSandbox) {
      internalPlugins.push(lovableBuildErrorDiagnostics());
    }
    const { tanstackStart } = await import("@tanstack/react-start/plugin/vite");
    const tanstackStartDefaults = {
      importProtection: {
        behavior: "error",
        client: {
          files: ["**/server/**"],
          specifiers: ["server-only"]
        }
      }
    };
    const tanstackStartOptions = mergeConfig(tanstackStartDefaults, options.tanstackStart ?? {});
    internalPlugins.push(tanstackStart(tanstackStartOptions));
    const explicitNitro = options.nitro === true || typeof options.nitro === "object" && options.nitro !== null;
    const shouldRunNitro = options.nitro !== false && command === "build";
    if (shouldRunNitro) {
      let nitroMod;
      try {
        nitroMod = await import("nitro/vite");
      } catch (err) {
        if (explicitNitro || isSandbox) {
          const reason = explicitNitro ? `an explicit \`nitro\` option (${options.nitro === true ? "true" : "{ ... }"}) was set in vite.config.ts` : "running inside a Lovable sandbox";
          throw new Error(
            `[@lovable.dev/vite-tanstack-config] ${reason}, but the \`nitro\` package isn't installed. Add \`nitro\` to your devDependencies (matches peerDependency >=3.0.260429-beta). Original import error: ${err instanceof Error ? err.message : String(err)}`
          );
        }
      }
      if (nitroMod) {
        const { nitro } = nitroMod;
        const userNitroOpts = typeof options.nitro === "object" && options.nitro ? options.nitro : {};
        const nitroOpts = { defaultPreset: "cloudflare-module", ...userNitroOpts };
        if (isSandbox) {
          delete nitroOpts.defaultPreset;
          nitroOpts.preset = "cloudflare-module";
          nitroOpts.output = { dir: "dist", serverDir: "dist/server", publicDir: "dist/client" };
          nitroOpts.cloudflare = { nodeCompat: true, deployConfig: true, ...userNitroOpts.cloudflare };
        } else if (!nitroOpts.preset && !await nitroSupportsDefaultPreset()) {
          logWarning(
            'The installed `nitro` predates the `defaultPreset` option (3.0.260603-beta), so the Cloudflare fallback is ignored and a zero-config build may target Node instead. Upgrade nitro to >=3.0.260603-beta, or set an explicit `nitro: { preset: "cloudflare-module" }`.'
          );
        }
        internalPlugins.push(nitro(nitroOpts));
      }
    }
    const viteReact = (await import("@vitejs/plugin-react")).default;
    internalPlugins.push(viteReact(options.react));
    if (command === "serve") {
      const hmrGateOpt = options.hmrGate;
      const enableHmrGate = isSandbox ? hmrGateOpt !== false : !!hmrGateOpt;
      if (enableHmrGate) {
        const { hmrGatePlugin } = await import("@lovable.dev/vite-plugin-hmr-gate");
        const hmrGateOptions = typeof hmrGateOpt === "object" ? hmrGateOpt : {};
        internalPlugins.push(hmrGatePlugin(hmrGateOptions));
      }
    }
    if (command === "serve" && isSandbox) {
      const { devServerBridgePlugin } = await import("@lovable.dev/vite-plugin-dev-server-bridge");
      internalPlugins.push(devServerBridgePlugin());
    }
    if (command === "serve") {
      internalPlugins.push(lovableAssetsProxyPlugin());
    }
    if (command === "serve" && mode === "development") {
      internalPlugins.push(componentTagger({ jsxSource: true }));
    }
    let envDefine = {};
    if (options.envDefine !== false) {
      const loadedEnv = loadEnv(mode, process.cwd(), "VITE_");
      for (const [key, value] of Object.entries(loadedEnv)) {
        envDefine[`import.meta.env.${key}`] = JSON.stringify(value);
      }
    }
    let config = {
      define: envDefine,
      // Client-scoped so React DevTools gets the dev react-dom; a global NODE_ENV
      // flip would emit jsxDEV, which the react-server SSR runtime can't resolve.
      ...isDevBuild ? {
        environments: { client: { define: { "process.env.NODE_ENV": JSON.stringify("development") } } },
        esbuild: { keepNames: true }
      } : {},
      // Match the build's CSS pipeline in dev. Vite uses PostCSS in dev and only
      // runs Lightning CSS at build, so build-time transforms (e.g. collapsing a
      // hand-written `-webkit-backdrop-filter` to the prefixed form Chrome ignores)
      // break the built/static output while the dev preview looks fine. Running
      // Lightning CSS in both keeps the preview honest.
      css: { transformer: "lightningcss" },
      resolve: {
        alias: {
          "@": `${process.cwd()}/src`
        },
        dedupe: [
          "react",
          "react-dom",
          "react/jsx-runtime",
          "react/jsx-dev-runtime",
          "@tanstack/react-query",
          "@tanstack/query-core"
        ]
      },
      // Dep re-optimization rotates the optimized-dep hash and 504s tabs holding
      // the old one; pre-bundle the always-present client deps + tolerate stale
      // requests. React core only — including @tanstack/react-start would pull its
      // node:async_hooks server entry into the client bundle and crash hydration.
      optimizeDeps: {
        include: ["react", "react-dom", "react-dom/client", "react/jsx-runtime", "react/jsx-dev-runtime"],
        ignoreOutdatedRequests: true
      },
      plugins: [...internalPlugins, ...options.plugins ?? []]
    };
    if (options.vite) {
      config = mergeConfig(config, options.vite);
    }
    if (isSandbox) {
      validateConfig(config);
      config = cleanServerConfig(
        mergeConfig(config, {
          server: { host: "::", port: 8080, strictPort: true, watch: { ignored: [...TOOL_MANAGED_WATCH_IGNORES] } }
        })
      );
    } else {
      config = mergeConfig({ server: { host: "::", port: 8080 } }, config);
    }
    if (isSandbox) {
      return config;
    }
    return applyWatchDebounceDefaults(config);
  };
}
export {
  defineConfig,
  lovableAssetsProxyPlugin
};
