// @lovable.dev/vite-plugin-dev-server-bridge/dist/index.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/index.ts
var src_exports = {};
__export(src_exports, {
  devServerBridgePlugin: () => devServerBridgePlugin
});
module.exports = __toCommonJS(src_exports);
var RESTART_ANNOUNCE_PATH = "/__lovable/restart-announce";
var RESTART_EVENT = "lovable:dev-server-restarting";
var PREVIEW_EXECUTE_PATH = "/_sandbox/preview/execute";
var PREVIEW_EXECUTE_EVENT = "lovable:preview-execute";
var PREVIEW_EXECUTE_RESULT_EVENT = "lovable:preview-execute-result";
var PREVIEW_EXECUTE_CLIENT_MODULE_ID = "virtual:lovable-preview-execute-client";
var RESOLVED_PREVIEW_EXECUTE_CLIENT_MODULE_ID = "\0" + PREVIEW_EXECUTE_CLIENT_MODULE_ID;
var RESTART_MAX_BODY_BYTES = 1024;
var PREVIEW_EXECUTE_MAX_BODY_BYTES = 64 * 1024;
var PREVIEW_EXECUTE_MAX_CODE_LENGTH = 50 * 1024;
var PREVIEW_EXECUTE_DEFAULT_TIMEOUT_MS = 1500;
var PREVIEW_EXECUTE_MAX_TIMEOUT_MS = 5e3;
var requestSequence = 0;
function nextRequestID() {
  requestSequence += 1;
  return `preview-exec-${Date.now().toString(36)}-${requestSequence.toString(36)}`;
}
function readJsonBody(req, maxBodyBytes) {
  return new Promise((resolve) => {
    let received = 0;
    const chunks = [];
    let aborted = false;
    req.on("data", (chunk) => {
      if (aborted)
        return;
      received += chunk.length;
      if (received > maxBodyBytes) {
        aborted = true;
        resolve({ ok: false, reason: "too_large" });
        return;
      }
      chunks.push(chunk);
    });
    req.on("end", () => {
      if (aborted)
        return;
      if (chunks.length === 0) {
        resolve({ ok: false, reason: "empty" });
        return;
      }
      try {
        resolve({ ok: true, value: JSON.parse(Buffer.concat(chunks).toString("utf8")) });
      } catch {
        resolve({ ok: false, reason: "invalid_json" });
      }
    });
    req.on("error", () => resolve({ ok: false, reason: "read_error" }));
  });
}
function writeJson(res, statusCode, value) {
  res.writeHead(statusCode, { "content-type": "application/json; charset=utf-8" });
  res.end(JSON.stringify(value));
}
function clampTimeoutMs(value) {
  if (typeof value !== "number" || !Number.isFinite(value))
    return PREVIEW_EXECUTE_DEFAULT_TIMEOUT_MS;
  return Math.max(1, Math.min(PREVIEW_EXECUTE_MAX_TIMEOUT_MS, Math.trunc(value)));
}
function parsePreviewExecuteRequest(value) {
  if (!value || typeof value !== "object")
    return "body must be a JSON object";
  const code = value.code;
  if (typeof code !== "string" || code.trim() === "")
    return "code must be a non-empty string";
  if (code.length > PREVIEW_EXECUTE_MAX_CODE_LENGTH) {
    return `code must be at most ${PREVIEW_EXECUTE_MAX_CODE_LENGTH} characters`;
  }
  return {
    code,
    timeout_ms: clampTimeoutMs(value.timeout_ms)
  };
}
function isPreviewExecuteResult(value) {
  if (!value || typeof value !== "object")
    return false;
  const result = value;
  return typeof result.request_id === "string" && typeof result.viewer_id === "string" && typeof result.ok === "boolean" && typeof result.duration_ms === "number";
}
function executeInViewers(ws, request, pendingRequests) {
  const requestID = nextRequestID();
  const viewerCount = ws.clients?.size ?? 0;
  const results = /* @__PURE__ */ new Map();
  pendingRequests.set(requestID, results);
  if (viewerCount === 0) {
    pendingRequests.delete(requestID);
    return Promise.resolve({
      request_id: requestID,
      viewer_count: 0,
      responded_count: 0,
      timed_out_count: 0,
      results: []
    });
  }
  ws.send({
    type: "custom",
    event: PREVIEW_EXECUTE_EVENT,
    data: {
      request_id: requestID,
      code: request.code
    }
  });
  return new Promise((resolve) => {
    let done = false;
    let timeout;
    let checkComplete;
    const finish = () => {
      if (done)
        return;
      done = true;
      clearTimeout(timeout);
      clearInterval(checkComplete);
      pendingRequests.delete(requestID);
      resolve({
        request_id: requestID,
        viewer_count: viewerCount,
        responded_count: results.size,
        timed_out_count: Math.max(0, viewerCount - results.size),
        results: [...results.values()]
      });
    };
    timeout = setTimeout(finish, request.timeout_ms);
    checkComplete = setInterval(() => {
      if (results.size >= viewerCount)
        finish();
    }, 10);
  });
}
function previewExecuteClientScript() {
  return `
if (import.meta.hot) {
  const viewerID = globalThis.crypto?.randomUUID?.() ?? Math.random().toString(36).slice(2);
  const handled = new Set();
  const AsyncFunction = Object.getPrototypeOf(async function () {}).constructor;
  const serialize = (value) => {
    if (value === undefined) return { type: "undefined" };
    try {
      return structuredClone(value);
    } catch {
      try {
        return JSON.parse(JSON.stringify(value));
      } catch {
        return String(value);
      }
    }
  };
  import.meta.hot.on("${PREVIEW_EXECUTE_EVENT}", async (payload) => {
    const started = performance.now();
    const requestID = payload && typeof payload.request_id === "string" ? payload.request_id : "";
    const code = payload && typeof payload.code === "string" ? payload.code : "";
    if (!requestID || !code || handled.has(requestID)) return;
    handled.add(requestID);
    try {
      const value = await new AsyncFunction(code)();
      import.meta.hot.send("${PREVIEW_EXECUTE_RESULT_EVENT}", {
        request_id: requestID,
        viewer_id: viewerID,
        ok: true,
        value: serialize(value),
        duration_ms: Math.round(performance.now() - started),
      });
    } catch (err) {
      import.meta.hot.send("${PREVIEW_EXECUTE_RESULT_EVENT}", {
        request_id: requestID,
        viewer_id: viewerID,
        ok: false,
        error: {
          name: err && typeof err === "object" && "name" in err ? String(err.name) : undefined,
          message: err instanceof Error ? err.message : String(err),
          stack: err instanceof Error ? err.stack : undefined,
        },
        duration_ms: Math.round(performance.now() - started),
      });
    }
  });
}
`;
}
function isPreviewExecuteClientCarrier(id) {
  return id.includes("/vite/dist/client/client.mjs");
}
function devServerBridgePlugin() {
  return {
    name: "lovable-dev-server-bridge",
    apply: "serve",
    configureServer(srv) {
      const pendingPreviewExecuteRequests = /* @__PURE__ */ new Map();
      const ws = srv.ws;
      ws.on?.(PREVIEW_EXECUTE_RESULT_EVENT, (data, client) => {
        if (!isPreviewExecuteResult(data))
          return;
        const results = pendingPreviewExecuteRequests.get(data.request_id);
        if (!results)
          return;
        results.set(client ?? data.viewer_id, data);
      });
      srv.middlewares.use((req, res, next) => {
        if (req.method !== "POST")
          return next();
        const path = (req.url ?? "").split("?", 1)[0];
        if (path !== RESTART_ANNOUNCE_PATH && path !== PREVIEW_EXECUTE_PATH)
          return next();
        const maxBodyBytes = path === PREVIEW_EXECUTE_PATH ? PREVIEW_EXECUTE_MAX_BODY_BYTES : RESTART_MAX_BODY_BYTES;
        void readJsonBody(req, maxBodyBytes).then((parsed) => {
          if (path === PREVIEW_EXECUTE_PATH) {
            if (!parsed.ok) {
              writeJson(res, parsed.reason === "too_large" ? 413 : 400, { error: parsed.reason });
              return;
            }
            const previewRequest = parsePreviewExecuteRequest(parsed.value);
            if (typeof previewRequest === "string") {
              writeJson(res, 400, { error: previewRequest });
              return;
            }
            void executeInViewers(ws, previewRequest, pendingPreviewExecuteRequests).then((response) => {
              writeJson(res, 200, response);
            });
            return;
          }
          const reason = parsed.ok && parsed.value && typeof parsed.value === "object" && "reason" in parsed.value && typeof parsed.value.reason === "string" ? parsed.value.reason : void 0;
          srv.ws.send({
            type: "custom",
            event: RESTART_EVENT,
            data: reason ? { reason } : {}
          });
          res.writeHead(204);
          res.end();
        });
      });
    },
    resolveId(id) {
      if (id === PREVIEW_EXECUTE_CLIENT_MODULE_ID)
        return RESOLVED_PREVIEW_EXECUTE_CLIENT_MODULE_ID;
      return null;
    },
    load(id) {
      if (id === RESOLVED_PREVIEW_EXECUTE_CLIENT_MODULE_ID)
        return previewExecuteClientScript();
      return null;
    },
    transform(code, id, options) {
      if (options?.ssr)
        return null;
      if (!isPreviewExecuteClientCarrier(id))
        return null;
      return { code: `${code}
import(${JSON.stringify(PREVIEW_EXECUTE_CLIENT_MODULE_ID)});
`, map: null };
    }
  };
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  devServerBridgePlugin
});
